### ISSUE TITLE
MOCK PA2 Level 2 Stream

### ISSUE THREAD

**Issue by kaiju10** [nus-cs2030/2324-s2#572]
# Summary
Trying to use stream to return a Num for add()

## Description 
I understand that there are two ways for Level 2, which are for loops and Stream. However, I want to try out using the Stream. I kinda know that you will need iterate(), limit() and reduce(), but i am unable to generate a Num in the end due to logic errors when piecing them together. findFirst() will not work as well, as it will return an Optional<Num>, and it will be hard to deal with. Any tips that can nudge me in the correct direction?

**Comment by samuelneo** [comment]
instead of `limit`, note that `iterate` can take in (as its second argument) a Predicate that determines whether to continue iterating. you may want to refer to the Java 17 API on Streams for this

**Comment by ZYZY2727** [comment]
can use predicate hasNext = x -> (!x.equals(n)) into the stream iterate
then reduce it

**Comment by kaiju10** [comment]
okay, i can see how iterate() is being used to create a for loop. But how do i connect the reduce to the iterate()?

**Comment by xiang-qiuyu** [comment]
> okay, i can see how iterate() is being used to create a for loop. But how do i connect the reduce to the iterate()?

i am guessing that you are planning to use reduce to retrieve the last element by using reduce(this, (x, y) -> y)? can just add behind your iterate() since iterate returns a stream and reduce takes in a stream. hope this helps!

**Comment by kaiju10** [comment]
@xiang-qiuyu, yes it does help. I have been thinking of a way to extract the last element out. In the end the add() method works, but it is a variation:
1) First stream will generate the int version of the count. So it goes through `iterate()` with `zero()` as base, and then use `count()`. 
2) the int will be applied in the second stream, where it goes through `iterate()` with `this` as base, `limit(int)`, and `reduce()` to find last element.

i cant get it in a single stream, as i know the predicate is meant for `zero()` and not `this`, so it is hard to increment both within a single stream. I believe I am missing out on more functions for the `iterate()`. If you or anyone can advise on this, it will be much appreciated =)

**Comment by juliannxe** [comment]
have you considered using other methods instead of a stream? a while loop could also be possible

**Comment by xiang-qiuyu** [comment]
> @xiang-qiuyu, yes it does help. I have been thinking of a way to extract the last element out. In the end the add() method works, but it is a variation:
> 
> 1. First stream will generate the int version of the count. So it goes through `iterate()` with `zero()` as base, and then use `count()`.
> 2. the int will be applied in the second stream, where it goes through `iterate()` with `this` as base, `limit(int)`, and `reduce()` to find last element.
> 
> i cant get it in a single stream, as i know the predicate is meant for `zero()` and not `this`, so it is hard to increment both within a single stream. I believe I am missing out on more functions for the `iterate()`. If you or anyone can advise on this, it will be much appreciated =)

ahh okay i have trouble connecting the two streams also HAHA my bad, i misinterpreted your question

**Comment by kaiju10** [comment]
@juliannxe i did considered a while loop afterwards, but it will be better to deal with a stream so as to better prepare for PE. Thanks for your suggestion

@xiang-qiuyu i see, its alright, but still thx alot for ur input for reduce()

Thanks also to @ZYZY2727 and @samuelneo for helping me out. I will be closing the issue
